### PR TITLE
Fixes for aspect ratio and min/max sizes

### LINKED ISSUE #167: Reduce number of `Vec`s allocated in `perform_absolute_layout_on_absolute_children`
## What problem does this solve or what need does it fill?

Reduce number of `Vec`s generated in `perform_absolute_layout_on_absolute_children` in the flexbox algorithm:
<https://github.com/DioxusLabs/taffy/blob/e75a55f57946eeffb1a436925c88213c1102e57f/src/flexbox.rs#L1404-L1410>

## What solution would you like?

Whatever works best regarding the benchmarks.

## What alternative(s) have you considered?

We can leave it as-is.

## Additional context

This should be implemented (or at least merged) after #128 such that we can assess whether the changes actually improve performance.

### LINKED ISSUE #313: Support aspect-ratio for Leaf nodes
## What problem does this solve or what need does it fill?

The css aspect ratio property isn't supposed for leaf nodes.

## What solution would you like?

The support for that property out of the box
Also it would be nice to test if this property works well for grid/ flex nodes (I didn't find any test testing the property)

## What alternative(s) have you considered?

- Entering cross axis size manually
- Implementing this by myself, unfortunately the knowDimensions.width that passes into Leaf's compute function is equal to 100f32 (more context in the gist below) which is the parent's width (does this make sense?). Getting cross axis would be easy if that parameter was equal to Option::None. Also this is the first rust project i'm working with, I don't feel very confident in what I am doing yet.

## Additional context

I've done some researches to ensure this wasn't supported
Here is an adaptation of the test script (js) to take aspectRatio into account aswell as a simple test that shows the issue (rs & html files)
https://gist.github.com/adjabaev/4dfd194ddee6becab49c9dbdbd191baa

### PR DESCRIPTION
**Pull request by nicoburns**
# Objective

Fix #313 (and more generally fix support for aspect-ratio).
This PR now also fixes #167

## Subtasks

- [x] Aspect ratio support for leaf nodes
- [x] Aspect ratio support for grid containers
- [x] Aspect ratio support for grid items
- [x] Aspect ratio support for flex containers
- [x] Aspect ratio support for regular flex items
- [x] Aspect ratio support for absolutely positioned flex items
- [x] Min/max size support for grid items
- [x] Fix margins for absolutely positioned flex items
- [x] Fix min/max interactions with aspect ratio for absolutely positioned flex items
- [x] Fix min/max interaction with inset-generated dimensions for absolutely positioned flex items

## Context

Taffy has an `aspect_ratio` style property and claims to support CSS `aspect-ratio`. But actual support is patchy, and it is not covered by gentests.

### CODE REVIEW

**Comment by nicoburns** [comment]
I'm not sure if this covers everything (in fact I have 4 disabled tests), but aspect ratio is working a lot better with this PR than it was before (including the use case in #313), and I'm intending to move on to other tasks now and return to aspect ratio again later if required.

**Comment by adjabaev** [comment]
@nicoburns hey!
thank for the quick answer/ implementation!
I just started playing with it and noticed this test isn't passing:
(the inset doesn't seem to change anything, it is just a representation of a real use case)
```
// aspect_ratio_leaf_inset_width
<div id="test-root" style="display: flex; width: 1280px; height: 720px;">
  <div style="position: absolute; top: 5%; left: 5%; width: 20%; aspect-ratio: 3;"></div>
</div>
```

**Comment by nicoburns** [comment]
@adjabaev Well spotted! I've fixed that case, but I think I'll need to test more cases using absolute positioning.

**Comment by adjabaev** [comment]
https://github.com/DioxusLabs/taffy/pull/317/commits/86509aa0d4b3dc99f5a6c44f132e8f71f7353660

It might be a mistake but both tests you commited have identical content

On a side note, I'll keep trying getting crazy with aspect ratio and keep you updated if I end up finding something behaving weirdly

**Comment by nicoburns** [comment]
> It might be a mistake but both tests you commited (fixtures) have identical content

Ah yes, that'll be because the "rename" and "duplicate" buttons are right next to each other in my editor. I wanted to rename it "absolute" because it turns out that percentages have nothing to do with it.

**Comment by nicoburns** [comment]
This has turned up a few bugs in the final positioning code for grid items and some more in the final positioning code for absolutely positioned flex items. Around aspect ratio, but also around min/max width and margins.

None of this is that hard to fix, but it requires a *lot* of tests to test all the permutations because:

- it applies to min/max sizes and preferred sizes
- it applies to sizes generated by the inset properties
- it interacts with stretch alignment
- it applies differently with width and height because in many cases width is prioritised if an aspect ratio is set
- it is implemented separately for grid items, regular flex items and absolutely positioned flex items (requiring duplicate tests for each implementation)

So this is going to take a bit of grinding through all the test cases.

**Comment by alice-i-cecile** [comment]
I'm very glad to see these bugs slowly get resolved.

> it is implemented separately for grid items, regular flex items and absolutely positioned flex items (requiring duplicate tests for each implementation)

Should we try to deduplicate this?

File: src/compute/leaf.rs [hunk 1]
```diff
@@ -25,19 +26,21 @@ pub(crate) fn compute(
...
     // Resolve node's preferred/min/max sizes (width/heights) against the available space (percentages resolve to pixel values)
     // For ContentSize mode, we pretend that the node has no size styles as these should be ignored.
-    let (node_size, node_min_size, node_max_size) = match sizing_mode {
+    let (node_size, node_min_size, node_max_size, aspect_ratio) = match sizing_mode {
         SizingMode::ContentSize => {
             let node_size = known_dimensions;
             let node_min_size = Size::NONE;
             let node_max_size = Size::NONE;
-            (node_size, node_min_size, node_max_size)
+            (node_size, node_min_size, node_max_size, None)
```
Comment (Weibye) [inline src/compute/leaf.rs:34]: So ContentSize will not apply aspect ratio but always keep to the size of the content? 

What will be the correct approach for a user to display an image in a node, but wanting to define what aspect ratio to use? Would it involve setting ContentSize on the image, and setting aspect ratio on the parent node instead?

**Review by Weibye — COMMENTED** [review]
Some nitpicks and questions :)

File: src/geometry.rs [hunk 2]
```diff
@@ -365,6 +365,22 @@ impl Size<Option<f32>> {
...
+    ///
+    /// If aspect_ratio is `None` then this function simply returns self.
+    pub fn maybe_apply_aspect_ratio(self, aspect_ratio: Option<f32>) -> Size<Option<f32>> {
+        match aspect_ratio {
+            Some(ratio) => match (self.width, self.height) {
+                (Some(width), None) => Size { width: Some(width), height: Some(width / ratio) },
+                (None, Some(height)) => Size { width: Some(height * ratio), height: Some(height) },
+                _ => self,
+            },
+            None => self,
```
[the remark this answers is not in the document]
Comment (nicoburns) [reply src/geometry.rs:387]: For 2 you are correct: setting both and `height` and `width` properties overrides aspect ratio. CSS considers this valid, but then CSS considers any combination of properties valid and just ignores properties that don't make. I'd like to provide better diagnostics and warnings, but I'm convinced console warnings are the best way to do this. I think we ought to have a "warnings and diagnostics API" that allow integrators to introspect Taffy and build something akin to the browser dev tools.

[the remark this answers is not in the document]
Comment (nicoburns) [reply src/geometry.rs:387]: Case 1 depends on context. If there is something else that implicitly set width or height (for example the inset properties or stretch alignment) then the other dimension will be filled in via aspect ratio. If they are *both* set implicitly in this manner than width overrides height and the height is set via the aspect ratio.

File: src/compute/leaf.rs (hunk 1 above)
Comment (nicoburns) [reply src/compute/leaf.rs:34]: An image typically has an intrinsic aspect ratio (due to the ratio of the dimensions of the image file itself). But that would be out of scope for Taffy which doesn't know anything about images. The MeasureFunc for the image should apply this aspect ratio itself if it wants to. If the user wants to override this apsect ratio then they could set both a width (or height) and an aspect ratio on the leaf node and Taffy would apply this.

Medium term I suspect image nodes won't use Leaf layout with MeasureFuncs. Instead they will be their own custom layout modes (and responsible for managing their own styles outside of Taffy).

Comment (Weibye) [reply src/compute/leaf.rs:34]: Alright, thanks for the insight!

File: src/compute/flexbox.rs [hunk 3]
```diff
@@ -221,6 +221,11 @@ fn compute_preliminary(
     }
 
     // TODO: Add step 4 according to spec: https://www.w3.org/TR/css-flexbox-1/#algo-main-container
```
Comment (Weibye) [inline src/compute/flexbox.rs:223]: Should this TODO now be removed?

Comment (nicoburns) [reply src/compute/flexbox.rs]: Indeed. I thought I'd got that one!

Comment (nicoburns) [reply src/compute/flexbox.rs]: Indeed. I thought I'd got that one!

**Comment by nicoburns** [comment]
@alice-i-cecile This is now ready for another round of review / merging when ready.

File: src/compute/flexbox.rs [hunk 4]
```diff
@@ -1579,157 +1586,183 @@ fn final_layout_pass(tree: &mut impl LayoutTree, node: Node, flex_lines: &mut [F
...
-                        constants.padding_border.cross_start(constants.dir)
-                    }
+                (AlignSelf::Center, _) => {
+                    (constants.container_size.cross(constants.dir)
+                        + constants.padding_border.cross_start(constants.dir)
+                        - constants.padding_border.cross_end(constants.dir)
+                        - final_size.cross(constants.dir)
+                        + resolved_margin.cross_start(constants.dir)
+                        - resolved_margin.cross_end(constants.dir))
+                        / 2.0
```
Comment (Weibye) [inline src/compute/flexbox.rs:1758]: I particularly like this way of structuring this code!

**Review by Weibye — APPROVED** [review]
Looks good, I haven't been able to spot any issues but as usual I'm relying on your expertise and the tests to vouch for the correctness of the implementation.

File: src/geometry.rs [hunk 5]
```diff
@@ -42,6 +43,19 @@ impl<T: Default> Default for Rect<T> {
     }
 }
 
+impl<U, T: Add<U>> Add<Rect<U>> for Rect<T> {
```
Comment (alice-i-cecile) [inline src/geometry.rs:46]: This deserves a note in the release notes; it's surprisingly useful.

**Review by alice-i-cecile — APPROVED** [review]

**Pull request merged**